RFCs/0017-TestCase-FullyQualifiedName.md

```diff
@@ -1,49 +1,97 @@
+* Each TestCase has an `ID` field that takes the form of a GUID. In the past, this field has been optional. It will now become required. The required `ID` field must be unique for a given test assembly. This `ID` field must also be deterministic, in the sense that a test with the same class, method and method arguments should have the same id. We envision generating this `ID` by hashing the test's class, method, method signature and method arguments into the GUID. 
+* `FullyQualifiedName` should be immutable in that the test store keeps the exact value that is provided by the test adapter and supplies it back to the test adapter for test execution. The test adapter should be able to rely on receiving the same FQN for execution that it supplied when performing discovery.
+* `FullyQualifiedName` should reference the semantic location of the method that defines the test. This means that the `FullyQualifiedName` references the method on the type where the test is discovered. In some situations, this could be different from the type where the method is defined. See later in this document for a more detailed discussion.
 
+## `FullyQualifiedName` specification
 
-The normalized form allows the test platform to reference groups of tests in a more general form without having to interpret the logic that the test adapter uses when constructing the fully qualified name. 
+We are standardizing the `FullyQualifiedName` property on a TestCase in the following way. The FQN will be parsed as a URI starting with the scheme value of `fqn`. 
+
+**Method Form**
+
+    fqn://clr/m/{type}/{method}
+
+`fqn://` - Any FQN that starts with the characters `fqn://` will be considered to follow the standardized format.
+
+`clr` – The first segment ("host" segment) specifies how to interpret the rest of the URI. The only valid value at the time of this writing is "clr" to reflect support for tests written in CLR managed languages. This portion can be expanded in the future to add support for non-managed languages such as Python, Javascript, C++, etc…
+
+`m | o` – `'m'` indicates a method form FQN, `'o'` indicates a custom test case identifier that the test explorer will not attempt to interpret.
+
+`{type}` - The `type` segment includes the fully specified type name in metadata format:
+
+    NamespaceA.NamespaceB.ClassName`1+InnerClass`2
+
+* The type name must be fully qualified (in the CLR sense), including its namespace. Any generic classes must also include an arity value using backtick notation (`# where # is the number of type arguments that the class requires).
+* Nested classes are appended with a '+' and must also include an arity if generic.
```

Comment:
You should be more specific about the arity of nested types.  I forget the details, but I believe that they don't behave the way one may expect in reflection.

Reply:
@Pilchie I haven't been able to find a scenario that doesn't work this way with regard to nested types. 

```csharp
public class A<T>
{
    // Named by reflection as A`1+B`1
    public class B<X> { }

    // Named by reflection as A`1+B`2
    public class B<Y, Z> { }
}
```

Is there another scenario we should be considering?

Reply:
I think the issue I had in mind is just that reflection considers "T" to be a type parameter of B in a lot of ways.  Probably not an issue for the name.

Reply:
True, that the nested classes include the template parameters of the containing class. So, nested class B<X> functions as if it were declared B<T, X> with a silent first template argument. We are numbering the template arguments including those that are received from the containing class. This is consistent with C# and the CLR metadata spec. I think we are on the same page here.

Reply:
I've added an example to demonstrate this.